src/services/Batcher.ts

```diff
@@ -0,0 +1,144 @@
+import { MaybePromise } from '@/types'
+
+export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
+export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
+export type BatchOptions = {
+  maxBatchSize?: number,
+  maxWait?: number,
+}
+
+type Timer = ReturnType<typeof setTimeout>
+type Resolve<V> = (value: V) => void
+type Reject = (reason?: unknown) => void
+type BatchQueueValue<V, R> = {
+  response: Promise<R>,
+  resolve: Resolve<R>,
+  reject: Reject,
+}
+
+type BatchQueue<V, R> = Map<V, BatchQueueValue<V, R>>
+
+export class Batcher<V, R> {
+  private readonly callback: BatchCallback<V, R>
+  private readonly options: BatchOptions
+  private readonly queue: BatchQueue<V, R> = new Map()
+  private timeout: Timer | null = null
+  private waitingSince: number | null = null
+
+  public constructor(callback: BatchCallback<V, R>, options: BatchOptions = {}) {
+    this.callback = callback
+    this.options = options
+  }
+
+  public batch(value: V): Promise<R> {
+    if (this.queue.has(value)) {
+      const { response } = this.queue.get(value)!
+
+      return response
+    }
+
+    let resolve!: Resolve<R>
+    let reject!: Reject
+
+    const response = new Promise<R>((...args) => {
+      [resolve, reject] = args
+    })
+
+    this.queue.set(value, {
+      response,
+      resolve,
+      reject,
+    })
+
+    this.requestProcessQueue()
+
+    return response
+  }
+
+  private requestProcessQueue(): void {
+    if (this.shouldProcessNow()) {
+      this.processQueue()
+      return
+    }
+
+    this.waitToProcessQueue()
+  }
+
+  private waitToProcessQueue(): void {
+    if (this.waitingSince === null) {
+      this.waitingSince = Date.now()
+    }
+
+    if (this.timeout) {
+      clearTimeout(this.timeout)
+    }
+
+    this.timeout = setTimeout(() => this.processQueue())
+  }
+
+  private shouldProcessNow(): boolean {
+    return this.maxBatchSizeReached() || this.maxWaitReached()
+  }
+
+  private maxBatchSizeReached(): boolean {
+    const { maxBatchSize = Infinity } = this.options
+
+    return this.queue.size >= maxBatchSize
+  }
+
+  private maxWaitReached(): boolean {
+    const { maxWait = Infinity } = this.options
+    const now = Date.now()
+    const since = this.waitingSince ?? 0
+
+    return now - since >= maxWait
+  }
+
+  private getBatchToProcess(): BatchQueue<V, R> {
+    const batch = new Map(this.queue)
+
+    this.queue.clear()
+    this.waitingSince = null
+
+    if (this.timeout) {
+      clearTimeout(this.timeout)
+    }
+
+    return batch
+  }
+
+  private async processQueue(): Promise<void> {
+    const batch = this.getBatchToProcess()
+    const values = Array.from(batch.keys())
+
+    try {
+      const response = await this.callback(values)
+
+      if (typeof response === 'function') {
+        return this.resolveBatchUsingLookup(batch, response)
+      }
```

Comment:
A read me is a great idea. But I'll explain briefly here. The return type for a `BatchCallback` is

```typescript
export type BatchCallbackLookup<V, R> = (value: V) => MaybePromise<R>
export type BatchCallback<V, R> = (values: V[]) => MaybePromise<Map<V, R>> | BatchCallbackLookup<V, R>
```

You can return `Map<V, R>` OR a function that accepts V and returns R. This is for flexibility in implementation. 

If a map is returned, its assumed that for every `V` there will be a key in the map with its `R` as the value. 

If a lookup function is returned. Its assumed that it can be called with `V` and `R` will be returned.